packages/serverpod/lib/src/server/log_manager/log_writers.dart

```diff
@@ -0,0 +1,215 @@
+    _openLogEntry = entry;
+  }
+
+  @override
+  Future<int> closeLog(SessionLogEntry entry) async {
+    var openLogEntry = _openLogEntry ?? entry;
+
+    await _logWriter.openLog(openLogEntry);
+
+    for (var query in _queries) {
+      await _logWriter.logQuery(query);
+    }
+
+    for (var logEntry in _logEntries) {
+      await _logWriter.logEntry(logEntry);
+    }
+
+    for (var message in _messages) {
+      await _logWriter.logMessage(message);
+    }
+
+    return _logWriter.closeLog(entry);
+  }
+
+  @override
+  Future<void> logEntry(LogEntry entry) async {
+    _logEntries.add(entry);
+  }
+
+  @override
+  Future<void> logMessage(MessageLogEntry entry) async {
+    _messages.add(entry);
+  }
+
+  @override
+  Future<void> logQuery(QueryLogEntry entry) async {
+    _queries.add(entry);
+  }
+}
+
+@internal
+class DatabaseLogWriter extends LogWriter {
+  final Session _session;
+  int? _sessionLogId;
+
+  DatabaseLogWriter(this._session);
+
+  @override
+  Future<void> logEntry(LogEntry entry) async {
+    var id = _sessionLogId;
+    if (id == null) throw StateError('The log has not been opened');
+
+    entry.sessionLogId = id;
+    await _databaseLog(_session, entry);
+  }
+
+  @override
+  Future<void> logMessage(MessageLogEntry entry) async {
+    var id = _sessionLogId;
+    if (id == null) throw StateError('The log has not been opened');
+
+    entry.sessionLogId = id;
+    await _databaseLog(_session, entry);
+  }
+
+  @override
+  Future<void> logQuery(QueryLogEntry entry) async {
+    var id = _sessionLogId;
+    if (id == null) throw StateError('The log has not been opened');
+
+    entry.sessionLogId = id;
+    await _databaseLog(_session, entry);
+  }
+
+  @override
+  Future<void> openLog(SessionLogEntry entry) async {
+    if (_sessionLogId != null) throw StateError('The log is already open');
+    _sessionLogId = await _databaseLog(_session, entry);
+  }
+
+  @override
+  Future<int> closeLog(SessionLogEntry entry) async {
+    var id = _sessionLogId;
+    if (id == null) {
+      id = await _databaseLog(_session, entry);
+    } else {
+      entry.id = id;
+      await _databaseLogUpdate(_session, entry);
+    }
+
+    _sessionLogId = null;
+    return id;
+  }
+
+  Future<void> _databaseLogUpdate<T extends TableRow>(
+    Session session,
+    T entry,
+  ) async {
+    var tempSession = await session.serverpod.createSession(
+      enableLogging: false,
+    );
+
+    await tempSession.db.updateRow<T>(entry);
+
+    await tempSession.close();
+  }
+
+  Future<int> _databaseLog<T extends TableRow>(Session session, T entry) async {
+    var tempSession = await session.serverpod.createSession(
+      enableLogging: false,
+    );
+
+    var result = await tempSession.db.insertRow<T>(entry);
+
+    await tempSession.close();
+
+    var id = result.id;
+
+    if (id == null) throw StateError('Failed to insert log entry');
+    return id;
+  }
+}
+
+@internal
+class StdOutLogWriter extends LogWriter {
```

Comment:
Question (for my knowledge): These are all `stdout` writes. Where do we log errors?

Reply:
Ahh true, the objects that we are logging contains the error if there is one, Here I think we can branch inside the log writer and write to stderr if the logged object contains an error.